src/T8code.jl

```diff
@@ -157,6 +159,84 @@ not a system-provided `t8code` installation. In this case, T8code.jl is not usab
 preferences_set_correctly() = !(_PREFERENCE_LIBT8 == "t8code_jll" &&
                                 MPIPreferences.binary == "system")
 
+# Minimal reference tracker holding active t8code related objects
+# created throughout the life time of a Julia session. t8code objects
+# should remove themselves from the tracker when they get finalized.
+if !@isdefined(T8CODE_OBJECT_TRACKER)
+    T8CODE_OBJECT_TRACKER = Dict()
+end
+
+"""
+    T8codeForestWrapper
+
+Lightweight `t8_forest_t` pointer wrapper which helps to free
+resources allocated by t8code in an orderly fashion.
+
+When initialized with a t8code forest pointer the wrapper
+registers itself with a t8code object tracker called `__T8CODE_OBJECT_TRACKER`.
+
+In serial mode the wrapper and in consequence the t8code forest
+can be finalized immediately whenever Julia's garbage collector sees fit.
+
+In (MPI) parallel mode the wrapper (and the t8code forest) is kept till the end
+of the session or when finalized explicitly. At the end of the session (resp.
+when the program shuts down) the object tracker finalizes all registered t8code
+objects in sync with all MPI ranks. This is necessary since t8code internally
+allocates MPI shared arrays. See `src/auxiliary/t8code.jl` for the finalization
+code when Trixi is shutting down.
+"""
+mutable struct ForestWrapper
+    pointer::Ptr{t8_forest} # cpointer to t8code forest
+    unique_id::UInt
+
+    function ForestWrapper(pointer::Ptr{t8_forest})
+        wrapper = new(pointer)
+
+        # Compute the unique id from the T8codeForestWrapper object.
+        wrapper.unique_id = pointer_from_objref(wrapper)
+
+        if MPI.Comm_size(MPI.Comm(t8_forest_get_mpicomm(pointer))) > 1
+            # Make sure the unique id is identical for each MPI rank.
+            wrapper.unique_id = MPI.bcast(wrapper.unique_id, MPI.Comm(t8_forest_get_mpicomm(pointer)))
+        end
+
+        finalizer(wrapper) do wrapper
+            # When finalizing, `forest`, `scheme`, `cmesh`, and `geometry` are
+            # also cleaned up from within `t8code`. The cleanup code for
+            # `cmesh` does some MPI calls for deallocating shared memory
+            # arrays. Due to garbage collection in Julia the order of shutdown
+            # is not deterministic. Hence, deterministic finalization is necessary in
+            # order to avoid MPI-related error output when closing the Julia
+            # program/session.
+            t8_forest_unref(Ref(wrapper.pointer))
+
+            # Deregister from the object tracker.
+            delete!(T8CODE_OBJECT_TRACKER, wrapper.unique_id)
+        end
+
+        # Register the T8codeForestWrapper with the object tracker.
+        T8CODE_OBJECT_TRACKER[wrapper.unique_id] = wrapper
```

Comment:
```suggestion
        T8CODE_OBJECT_TRACKER[wrapper.unique_id] = wrapper
        return wrapper
```
I was a bit confused at first as I expected the constructor to return a `ForestWrapper` until I realized that the last lines already implicitly returns `wrapper`.

Reply:
Your suggestion is more readable. I agree.